src/main/kotlin/com/openlattice/assembler/tasks/MaterializedEntitySetsDataRefreshTask.kt

```diff
@@ -0,0 +1,121 @@
+import com.openlattice.controllers.exceptions.ResourceNotFoundException
+import com.openlattice.edm.type.PropertyType
+import com.openlattice.organization.OrganizationEntitySetFlag
+import com.openlattice.organization.OrganizationPrincipal
+import com.openlattice.postgres.mapstores.MaterializedEntitySetMapStore
+import com.openlattice.tasks.HazelcastFixedRateTask
+import com.openlattice.tasks.Task
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+class MaterializedEntitySetsDataRefreshTask : HazelcastFixedRateTask<MaterializedEntitySetsDependencies> {
+    override fun getInitialDelay(): Long {
+        return 0L
+    }
+
+    override fun getPeriod(): Long {
+        return 3600L // minimum refresh rate is 1 min
+    }
+
+    override fun getTimeUnit(): TimeUnit {
+        return TimeUnit.MILLISECONDS
+    }
+
+    /**
+     * For every materialized entity set, if it's unsynchronized with data changes ( has a
+     * [com.openlattice.organization.OrganizationEntitySetFlag.DATA_UNSYNCHRONIZED] flag ) we check if the user set
+     * refresh rate time has passed since the last refresh. If yes, we refresh the materialized view of the entity set,
+     * if not, we skip it and will pick it up in the next period.
+     */
+    override fun runTask() {
+        val dataUnsynchPredicate = Predicates.equal(
+                MaterializedEntitySetMapStore.FLAGS_INDEX,
+                OrganizationEntitySetFlag.DATA_UNSYNCHRONIZED)
+
+        val hasRefreshRatePredicate = SqlPredicate("${MaterializedEntitySetMapStore.REFRESH_RATE_INDEX} != null")
+
+        val refreshableEntitySets = getDependency().materializedEntitySets
+                .aggregate(
+                        MaterializedEntitySetsRefreshAggregator(),
+                        Predicates.and(hasRefreshRatePredicate, dataUnsynchPredicate)
+                                as Predicate<EntitySetAssemblyKey, MaterializedEntitySet>)
+
+        refreshableEntitySets.forEach {
+            getDependency().assembler.refreshMaterializedEntitySet(
+                    it.organizationId,
+                    it.entitySetId,
+                    getAuthorizedPropertiesForMaterialization(it.organizationId, it.entitySetId))
+        }
+    }
+
+    private fun getAuthorizedPropertiesForMaterialization(organizationId: UUID, entitySetId: UUID): Map<UUID, PropertyType> {
+        val organizationPrincipal = getDependency().organizations.getOrganizationPrincipal(organizationId)
+        //This will be rare, since it is unlikely you have access to an organization that does not exist.
+                ?: throw ResourceNotFoundException("Organization does not exist.")
+
+        // check materialization on all linking and normal entity sets
+        val entitySet = getDependency().edm.getEntitySet(entitySetId)
+        val allEntitySetIds = mutableSetOf(entitySet.id) + entitySet.linkedEntitySets
+
+        allEntitySetIds.forEach { ensureMaterialize(it, organizationPrincipal) }
```

Comment:
If an organization has MATERIALIZE on a linked entity set but is missing permissions on one of its normal entity sets, we might still want to materialize the linked entity set without data from the entity set missing permissions instead of blocking materialization altogether (basically mimic read behavior).

Reply:
Right now, we only allow read, when the calling user has read on all normal entity sets, so this is mimicking it. I don't remember the whole conversation, but it has to do with not being able to filter values by user for materialized entity sets, because we don't know which value is from which normal entityset, so we only allow read, if user has read on all normal entity sets, hence do the same for materializing.

Reply:
Ah true we are checking read on all for reading linked entities -- I don't think necessarily want to be doing that. The partial permissions concern was for the case where you only have READ on first name in entity set A, and only READ on last name in entity set B, and linked A+B allows you to reconstruct full names. This was the rationale for requiring READ on all entity sets contributing data for a property type, and also isn't a concern if an entire entity set is filtered out.

(Requiring READ on all entity sets of a property type also doesn't actually prevent people from reconstructing names by doing things like creating associations with PII to normal entities involved in a linked entity set.)

Reply:
True, but still, after we materialize a linking entity set and we start to add read permissions on the postgres columns, we cannot filter out values for which the organization member has or has not read permissions for in the normal entity sets.
If we start to filter out whole normal entity sets from materialization (and loading linking entity sets), we will still require the organization member to have read on that whole sub-set of normal entity sets, which provided the data for that linking entity set and the org principal had materialized on.

Reply:
That makes sense -- theoretically we can compute user permissions per column based on the entity sets the organization is authorized to materialize, but that could get overly complex. I think this is fine for now but we might want to revisit the way linking permissions are granted

Reply:
I will include this on wiki page and and a TODO